flight/src/server.rs

```diff
@@ -0,0 +1,267 @@
+        &self,
+        highwater_mark: i64,
+        limit: i64,
+    ) -> anyhow::Result<Vec<ConclusionEvent>> {
+        self.as_ref()
+            .conclusion_events_since(highwater_mark, limit)
+            .await
+    }
+}
+
+struct FeedTable<T> {
+    feed: Arc<T>,
+    schema: SchemaRef,
+}
+
+impl<T> FeedTable<T> {
+    fn new(feed: Arc<T>) -> Self {
+        Self {
+            feed,
+            schema: Arc::new(Schema::new(vec![
+                Field::new("index", DataType::UInt64, false),
+                Field::new("event_type", DataType::UInt8, false),
+                Field::new("stream_cid", DataType::Binary, false),
+                Field::new("controller", DataType::Utf8, false),
+                Field::new("event_cid", DataType::Binary, false),
+                Field::new("data", DataType::Binary, true),
+                Field::new(
+                    "previous",
+                    DataType::List(Arc::new(Field::new("item", DataType::Binary, false))),
+                    true,
+                ),
+            ])),
+        }
+    }
+    fn highwater_mark_from_expr(expr: &Expr) -> Option<u64> {
+        let find_highwater_mark = |col: &Expr, lit: &Expr| {
+            col.try_as_col()
+                .map_or(false, |column| column.name == "index")
+                .then(|| {
+                    if let Expr::Literal(ScalarValue::UInt64(highwater_mark)) = lit {
+                        highwater_mark.to_owned()
+                    } else {
+                        None
+                    }
+                })
+                .flatten()
+        };
+        match expr {
+            Expr::BinaryExpr(expr) => match expr.op {
+                datafusion::logical_expr::Operator::Gt => {
+                    find_highwater_mark(expr.left.as_ref(), expr.right.as_ref())
+                }
+                datafusion::logical_expr::Operator::LtEq => {
+                    find_highwater_mark(expr.right.as_ref(), expr.left.as_ref())
+                }
+                _ => None,
+            },
+            _ => None,
+        }
+    }
+}
+#[async_trait::async_trait]
+impl<T: ConclusionFeed + std::fmt::Debug + 'static> TableProvider for FeedTable<T> {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+
+    fn table_type(&self) -> TableType {
+        TableType::Base
+    }
+    #[instrument(skip(self,_state), ret(level = Level::DEBUG))]
+    async fn scan(
+        &self,
+        _state: &dyn Session,
+        projection: Option<&Vec<usize>>,
+        filters: &[Expr],
+        limit: Option<usize>,
+    ) -> datafusion::common::Result<Arc<dyn ExecutionPlan>> {
+        let schema = projection
+            .map(|projection| self.schema.project(projection))
+            .transpose()?
+            .map(Arc::new)
+            .unwrap_or_else(|| self.schema.clone());
+        debug!(?schema, "projected schema");
+        Ok(Arc::new(FeedExec {
```

Comment:
nit : I see that the FeedExec implements the ExecutionPlan. Can we add comments as to when the execute gets invoked. Also some explanatory comments in the execute method might help. Why are we setting the limit to 10000 by default?

Reply:
I'll add some comments, but these type implement the traits from datafusion which are well commented. I recommend reading the TableProvider and ExecutionPlan trait comments.